### PR TITLE
Make the last two weeks of work available

### PR DESCRIPTION
**Pull request by rubenvdlinde**

### CODE REVIEW

File: lib/Service/AuthenticationService.php [hunk 1]
```diff
@@ -3,12 +3,27 @@
 namespace OCA\OpenConnector\Service;
```
Comment (rjzondervan) [inline lib/Service/AuthenticationService.php:3]: This class still needs docblocks @rjzondervan could you write those?

File: lib/Service/CallService.php [hunk 2]
```diff
@@ -15,21 +16,53 @@
...
+			}
+
+			return $value;
+	}
+	private function renderConfiguration(array $configuration, Source $source): array
+	{
+		$configuration = array_map(function($value) use ($source) { return $this->renderValue($value, $source);}, $configuration);
+
+		return $configuration;
 	}
```
Comment (rjzondervan) [inline lib/Service/CallService.php:66]: These functions still need docblocks @rjzondervan

**Review by rjzondervan — CHANGES_REQUESTED** [review]

File: lib/Controller/MappingsController.php [hunk 3]
```diff
@@ -162,4 +166,127 @@ public function destroy(int $id): JSONResponse
...
+    public function test(ObjectService $objectService, IURLGenerator $urlGenerator): JSONResponse
+    {
+		$openRegisters = $objectService->getOpenRegisters();
+
+        // Get all parameters from the request
+        $data = $this->request->getParams();
+
+
+        // Validate that required parameters are present
+        if (!isset($data['inputObject']) || !isset($data['mapping'])) {
```
Comment (WilcoLouwerse) [inline lib/Controller/MappingsController.php:215]: !isset
=== false

File: lib/Controller/MappingsController.php [hunk 4]
```diff
@@ -162,4 +166,127 @@ public function destroy(int $id): JSONResponse
...
+     *     "resultObject": {
+     *         "fullName": "John Doe",
+     *         "userAge": 30,
+     *         "contactEmail": "john@example.com"
+     *     },
+     *     "isValid": true,
+     *     "validationErrors": []
+     * }
+     */
+    public function test(ObjectService $objectService, IURLGenerator $urlGenerator): JSONResponse
```
Comment (WilcoLouwerse) [inline lib/Controller/MappingsController.php:206]: Function could use some cleanup / split up code

File: lib/Controller/MappingsController.php [hunk 5]
```diff
@@ -162,4 +166,127 @@ public function destroy(int $id): JSONResponse
...
+			}
+        }
+
+        // Check if validation is requested
+        if (empty($data['validation']) === false) {
+            $validation = $data['validation'];
+        }
+
+        // Create a new Mapping object and hydrate it with the provided mapping
+        $mappingObject = new Mapping();
```
Comment (WilcoLouwerse) [inline lib/Controller/MappingsController.php:255]: Why do we create a new Mapping object?
And if it its purpose is temporary, shouldn't we delete it after testing the mapping?

File: lib/Controller/MappingsController.php [hunk 6]
```diff
@@ -162,4 +166,127 @@ public function destroy(int $id): JSONResponse
...
+            // If mapping fails, return an error response
+            return new JSONResponse([
+                'error' => 'Mapping error',
+                'message' => $e->getMessage()
+            ], 400);
+        }
+
+        // Initialize validation variables
+        $isValid = true;
+        $validationErrors = [];
```
Comment (WilcoLouwerse) [inline lib/Controller/MappingsController.php:271]: You could combine these 2 by checking if $validationErrors is empty or not

File: lib/Service/AuthenticationService.php [hunk 7]
```diff
@@ -3,12 +3,27 @@
...
+use Jose\Component\Signature\Algorithm\RS384;
+use Jose\Component\Signature\Algorithm\RS512;
+use Jose\Component\Signature\JWSBuilder;
+use Jose\Component\Signature\Serializer\CompactSerializer;
 use OAuthException;
 use Symfony\Component\HttpFoundation\Exception\BadRequestException;
+use Twig\Environment;
+use Twig\Loader\ArrayLoader;
 
 class AuthenticationService
```
Comment (WilcoLouwerse) [inline lib/Service/AuthenticationService.php:24]: Please add class and function docblocks

File: lib/Service/CallService.php [hunk 8]
```diff
@@ -15,21 +16,53 @@
 use GuzzleHttp\Psr7\Response;
 use OCA\OpenConnector\Db\CallLog;
 use OCA\OpenConnector\Db\CallLogMapper;
+use OCA\OpenConnector\Twig\AuthenticationExtension;
+use OCA\OpenConnector\Twig\AuthenticationRuntimeLoader;
 use Symfony\Component\Uid\Uuid;
+use Twig\Environment;
+use Twig\Loader\ArrayLoader;
 
 class CallService
```
Comment (WilcoLouwerse) [inline lib/Service/CallService.php:25]: Please add class and function docblocks

**Review by remko48 — APPROVED** [review]
FE approved

File: lib/Twig/AuthenticationRuntime.php [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+
+class AuthenticationRuntime implements RuntimeExtensionInterface
+{
+	public function __construct(
+		private readonly AuthenticationService $authService,
+	) {
+
+	}
+
+	public function oauthToken(string $endpoint, Source $source): string
```
Comment (WilcoLouwerse) [inline lib/Twig/AuthenticationRuntime.php:17]: docblock

File: lib/Twig/AuthenticationRuntime.php [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+	{
+
+		return $this->authService->fetchOAuthTokens(
+			endpoint: $endpoint,
+			configuration: $source->getAuthenticationConfig()
+		);
+	}
+	public function jwtToken(Source $source): string
```
Comment (WilcoLouwerse) [inline lib/Twig/AuthenticationRuntime.php:25]: newline &
docblock

File: lib/Twig/MappingRuntime.php [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+class MappingRuntime implements RuntimeExtensionInterface
+{
+	public function __construct(
+		private readonly MappingService $mappingService,
+		private readonly MappingMapper  $mappingMapper
+	) {
+
+	}
+
+	public function executeMapping(Mapping|array|string|int $mapping, array $input, bool $list = false): array
```
Comment (WilcoLouwerse) [inline lib/Twig/MappingRuntime.php:21]: docblock

File: lib/Twig/MappingRuntime.php [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+	{
+		if(is_array($mapping) === true) {
```
Comment (WilcoLouwerse) [inline lib/Twig/MappingRuntime.php:23]: if (

**Review by WilcoLouwerse — CHANGES_REQUESTED** [review]

**Review by RalkeyOfficial — APPROVED** [review]
FE approved

File: lib/Controller/MappingsController.php (hunk 5 above)
Comment (rjzondervan) [reply lib/Controller/MappingsController.php:255]: It is never persisted to the database ;)

File: lib/Service/AuthenticationService.php (hunk 7 above)
Comment (WilcoLouwerse) [reply lib/Service/AuthenticationService.php:27]: Functions have docblocks, but class still doesn't

**Review by WilcoLouwerse — APPROVED** [review]

**Pull request merged**